tao_compiler/mlir/xla/ral/context/common_context_impl_mkldnn.cc

```diff
@@ -607,21 +581,117 @@ struct OnednnACLKeyHasher {
+  packed_weight_ = driver_->raw_alloc(ctx_, packed_weight_size);
+  cblas_sgemm_pack(
+      CblasRowMajor, CblasBMatrix, key.transpose_b ? CblasTrans : CblasNoTrans,
+      key.m, key.n, key.k, 1.0, static_cast<float*>(key.const_weight_ptr),
+      key.transpose_b ? key.k : key.n, static_cast<float*>(packed_weight_));
+}
+
+MklPackedWeight::~MklPackedWeight() {
+  if (!packed_weight_) return;
+  driver_->raw_dealloc(ctx_, packed_weight_);
+}
+
+using MklGemmCache =
+    ideep::utils::lru_cache<CpuGemmKey, std::shared_ptr<MklPackedWeight>,
+                            CpuGemmKeyMap>;
+
+struct MklGemmState : public Context::Resource {
+  std::mutex mu;
+  MklGemmCache cache{getWeightPrePackingCacheCapacity()};
+};
+
+static std::thread::id kMklDefaultThreadId;
+
+#endif
+
+template <typename Tinput, int N = 2, typename Tweight = Tinput,
+          typename Toutput = Tinput>
+void mkl_ral_gemm(ExecutionContext* ctx, void* stream_handle,
+                  MemRefType<Tinput, N> A, MemRefType<Tweight, N> B,
+                  MemRefType<Toutput, N> C, bool tp_a, bool tp_b,
+                  bool weight_is_const) {
+#if not defined(TAO_X86)
+  ctx->signalError(Context::FAILURE, "mkl_ral_gemm not impl");
+#else
+  int m = tp_a ? A.sizes[1] : A.sizes[0];
+  int k = tp_a ? A.sizes[0] : A.sizes[1];
+  int n = tp_b ? B.sizes[0] : B.sizes[1];
+
+  if (!enableWeightPrePacking() || !weight_is_const) {
+    cblas_sgemm(CblasRowMajor, tp_a ? CblasTrans : CblasNoTrans,
+                tp_b ? CblasTrans : CblasNoTrans, m, n, k, 1.0,
+                reinterpret_cast<Tinput*>(A.data), A.strides[0],
+                reinterpret_cast<Tweight*>(B.data), B.strides[0], 0.0,
+                reinterpret_cast<Toutput*>(C.data), C.strides[0]);
+    return;
+  }
+
+  std::string unique_name =
+      "tao_ral.cpu.mkl_gemm_" + tao::ral::TaoTypeNameHelper<Tinput>::Invoke();
+  auto state = ctx->getOrCreateResource<MklGemmState>(
+      unique_name, []() { return new MklGemmState; });
+  opaque_t packed_weight;
+  {
+    CpuGemmKey key{m, n, k, 1, tp_a, tp_b, B.data, kMklDefaultThreadId};
+    std::lock_guard<std::mutex> l(state->mu);
+    auto& cache = state->cache;
+    auto it = cache.find(key);
+    if (it == cache.end()) {
+      std::shared_ptr<MklPackedWeight> packed_weight_ptr(
+          new MklPackedWeight(ctx, key));
+      it = cache.insert(std::make_pair(key, packed_weight_ptr)).first;
+    }
+    packed_weight = it->second->packed_weight();
+  }
+  cblas_sgemm_compute(CblasRowMajor, tp_a ? CblasTrans : CblasNoTrans,
+                      CblasPacked, m, n, k, reinterpret_cast<Tinput*>(A.data),
+                      A.strides[0], reinterpret_cast<Tweight*>(packed_weight),
+                      B.strides[0], 0.0, reinterpret_cast<Toutput*>(C.data),
+                      C.strides[0]);
+#endif
+}
+
+struct OnednnGemmState : public Context::Resource {
+  std::mutex mu;
+  std::unordered_map<opaque_t, std::vector<ideep::tensor>> packed_weight_cache;
```

Comment:
Better to replace the `std::unordered_map<opaque_t, std::vector<ideep::tensor>>` with a LRU cache which takes `{weight_pointer, weight_desc}` as the key. This could be done in another PR.